lib/xgu/xgu.h

```diff
@@ -0,0 +1,535 @@
+                                  ((x2-1)<<16)|x1);
+}
+
+inline
+DWORD* xgu_set_clear_rect_vertical(DWORD* p, uint32_t y1, uint32_t y2) {
+    return push_command_parameter(p, NV097_SET_CLEAR_RECT_VERTICAL,
+                                  ((y2-1)<<16)|y1);
+}
+
+inline
+DWORD* xgu_set_clear_rect(DWORD* p,
+                          uint32_t x1, uint32_t x2,
+                          uint32_t y1, uint32_t y2) {
+    p = push_command(p, NV097_SET_CLEAR_RECT_HORIZONTAL, 2);
+    p = push_parameter(p, ((x2-1)<<16|x1));
+    return push_parameter(p, ((y2-1)<<16|y1));
+}
+
+inline
+DWORD* xgu_set_object(DWORD* p, uint32_t instance) {
+    return push_command_parameter(p, NV097_SET_OBJECT, instance);
+}
+
+inline
+DWORD* xgu_set_texgen_s(DWORD* p, XguTexgen tg) {
+    return push_command_parameter(p, NV097_SET_TEXGEN_S, tg);
+}
+
+inline
+DWORD* xgu_set_texture_matrix_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_TEXTURE_MATRIX_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_projection_matrix(DWORD* p, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_PROJECTION_MATRIX, m);
+}
+
+inline
+DWORD* xgu_set_model_view_matrix(DWORD* p, uint32_t bone_index, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_MODEL_VIEW_MATRIX + bone_index*4, m);
+}
+
+inline
+DWORD* xgu_set_inverse_model_view_matrix(DWORD* p, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_INVERSE_MODEL_VIEW_MATRIX, m);
+}
+
+inline
+DWORD* xgu_set_composite_matrix(DWORD* p, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_COMPOSITE_MATRIX, m);
+}
+
+inline
+DWORD* xgu_set_texture_matrix(DWORD* p, uint32_t slot, float m[4*4]) {
+    assert(slot == 0);
+    return push_command_matrix4x4(p, NV097_SET_TEXTURE_MATRIX, m);
+}
+
+/* ==== Stencil OP ==== */
+
+inline
+DWORD* xgu_set_stencil_op_fail(DWORD* p, XguStencilOp so) {
+    return push_command_parameter(p, NV097_SET_STENCIL_OP_fail, so);
+}
+
+inline
+DWORD* xgu_set_stencil_op_zfail(DWORD* p, XguStencilOp so) {
+    return push_command_parameter(p, NV097_SET_STENCIL_OP_zfail, so);
+}
+
+inline
+DWORD* xgu_set_stencil_op_zpass(DWORD* p, XguStencilOp so) {
+    return push_command_parameter(p, NV097_SET_STENCIL_OP_zpass, so);
+}
+
+/* ==== Vertex Data Array ==== */
+
+inline
+DWORD* xgu_set_vertex_data_array_format(DWORD* p, XguVertexArray index, XguVertexArrayType format,
+                                        uint32_t size, uint32_t stride) {
+    return push_command_parameter(p, NV097_SET_VERTEX_DATA_ARRAY_FORMAT + index*4,
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE, format) |
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_SIZE, size) |
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_STRIDE, stride));
+}
+
+inline
+DWORD* xgu_set_vertex_data_array_offset(DWORD* p, XguVertexArray index, const uint32_t data) {
```

Comment:
last argument should be `const void*` probably - it points to (physical) memory

Reply:
Oh. Makes sense.